src/common/options.cc

```diff
@@ -4163,6 +4163,10 @@ std::vector<Option> get_rgw_options() {
     .set_default(1 * 1024 * 1024)
     .set_description(""),
 
+    Option("rgw_max_attrs_num_in_req", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    .set_default(0)
+    .set_description("The maximum number of metadata items that can be put via single request"),
```

Comment:
I'm not sure where this goes, but there's a max limit on this value that's sized by civetweb.h mg_request_info http_headers[64].  The true limit is "somewhat less" than this number.  It would be good to document this number somewhere.  It would be good to emit a runtime warning at startup for values near this limit.  It would be really good if civetweb failed requests in a more obvious fashion when this limit was exceeded.  And it would be great if civetweb didn't have this limit in the first place (but given asio we might not want to bother.)

Reply:
Also, why "default(0)"?  ocata swift advertises 90 here (that means changing civetweb tho)

Reply:
Thanks for pointing out this CivetWeb's limitation! Documenting it is definitely a very good idea.

Reply:
> Also, why "default(0)"? ocata swift advertises 90 here (that means changing civetweb tho)

I believe that's one of the compromises coming from being a multi-API program. The `0` value means *do not enforce* as in `rgw_get_request_metadata` we have:
```cpp
      if (rgw_max_attrs_num_in_req &&
          ++valid_meta_count > rgw_max_attrs_num_in_req) {
        return -E2BIG;
      }
```

I see 2 factors driving this behavior:
* the need to not break the level of S3 compliance we offer in the default configuration,
* necessity to have the policy coherent between all APIs similarly to ACLs.

Reply:
But it's not true we have a '0 = do not enforce' limit.  What we currently have is a default "number slightly larger than 60 causes weird bad behavior" limit.  Having a common default that causes non-weird deterministic behavior in s3 would still be an improvement over what we offer today.

If it's really the case that s3 doesn't expect a limit here than perhaps this should be a swift-only limit.  And if we don't want to have any limit as a default, then the radosgw install directions need to call that out as something people caring about swift standards must set.  But before any of this makes sense, we'd have to fix the civetweb restriction which requires we have a limit.

Reply:
We have many different limits across the layers. Passing a check on particular layer doesn't guarantee success elsewhere.

I perceive CivetWeb as a component located on the front-end layer. As it can't deal properly with requests carrying huge number of metadata headers, it should at least properly error *there* (maybe by sending a `5xx`?). I would like to avoid introducing a logic that crosses the layer boundary.